developing/healthcheck/README.md

```diff
@@ -0,0 +1,44 @@
+
+Many apps will try to achieve their subset of the above goals via a single endpoint, but apps grow in sophistication it is not unheard of to tease those reponsibilities out to separate endpoints.
+
+## Authorization
+
+Frequently healthcheck endpoints are agnostic about authentication and authorization, meaning anyone that can reach the service via the network can access these endpoints. This implies that any information available in the healthcheck should not be confidential nor could be used to jeopardize the smooth running of the service.
+
+## Reliability
+
+Particular care should be taken when using healthchecks as part of the reliability story of an app.
+
+Prof von Neuman is credited with the concept of "synthesis of reliable organisms from unreliable components". He showed that the math works out as such: if my app is serially dependent on 3 other services that are each at 97% reliable, my app can be no more than ~91% (0.97 x 0.97 x 0.97) reliable. This means that if an app's healthcheck is setup to fail when any of its dependencies have failed, this has significantly limited the upper bound of the apps reliability.
+
+Sometimes an app's reliability really is limited by a dependency, e.g. a database connection. If every single endpoint and functionality in an app requires interaction with the app's database, then it is reasonable to call your app fully inoperable when the database connection is not functioning.
+
+However, if only a subset of the app's functionality is impacted by an unreliabile dependency, it may be better to handle that instability at runtime, rather than declaring the whole app down. Consider the following example: if only one (out of many) of an app's functionalities depend on sending email, the app should maybe not be considered wholly down if the email service is unreachable; it would be better to handle errors from interacting with the email service gracefully, or possibly even using a technique like feature flagging to turn off any attempts to use that functionality until the service is back in service.
+
+## Healthchecks in Deployment Pipelines
+
+It is not uncommon to see teams utilize an app's healthcheck endpoint on the critical path of a deployment pipeline, however this should considered for unintended consequences.
+
+In modern software engineering, automated deployments are a critical feature of a software system, facilitating the ability to safely and incrementally improve the quality of a running app.
+
+Again, consider the app example from above that uses an email sending service for one feature: should the external serivce instability to jeopardize the ability to roll out new code? Perhaps the new code for deployment is useful for diagnosing the issue, or is intended to temporarily turn off the offending sending of emails, blocking roll out of this code is actually making remediation of the issue harder rather than safer.
```

Comment:
```suggestion
Again, consider the app example from above that uses an email sending service for one feature: should the external service instability to jeopardize the ability to roll out new code? Perhaps the new code for deployment is useful for diagnosing the issue or is intended to temporarily turn off the offending sending of emails; blocking roll out makes remediation of the issue harder rather than safer.
```

Reply:
Thank you!